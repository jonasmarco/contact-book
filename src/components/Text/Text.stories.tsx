import { Meta, StoryFn } from '@storybook/react'
import styled, { css } from 'styled-components'

import Text, { Props } from '.'

const Background = styled.div`
  ${({ theme }) => css`
    background-color: ${theme.colors.tertiary};
    padding: 0;
  `};
`

export default {
  title: 'Text',
  component: Text,
  args: {
    children: 'Lorem ipsum'
  },
  argTypes: {
    children: {
      type: 'string'
    },
    danger: {
      type: 'boolean',
      defaultValue: false
    }
  },
  decorators: [
    (Story) => (
      <Background>
        <Story />
      </Background>
    )
  ],
  parameters: {
    layout: 'fullscreen',
    backgrounds: { default: 'dark' }
  }
} as Meta<typeof Text>

const Template: StoryFn<typeof Text> = (args: Props) => <Text {...args} />

export const NormalText = Template.bind({})
NormalText.args = {
  children: 'Lorem ipsum'
}

export const DangerText = Template.bind({})
DangerText.args = {
  children: "I'm dangerous baby!",
  danger: true
}
