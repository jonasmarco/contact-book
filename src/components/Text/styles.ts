import styled, { css } from 'styled-components'
import { customMedia } from '../MediaMatch'

import { Props } from '.'

type WrapperProps = Pick<Props, 'danger'>

export const Wrapper = styled.p<WrapperProps>`
  ${({ theme, danger }) => css`
    color: ${danger ? theme.colors.red : theme.colors.white};
    font-size: ${theme.fonts.sizes.small};
    line-height: ${theme.text.lineHeight};
    margin-bottom: ${theme.margin.textMargin};
    ${customMedia.greaterThan('tabletUp')`
      font-size: ${theme.fonts.sizes.normal};
    `};
    ${customMedia.greaterThan('notebookUp')`
      font-size: ${theme.fonts.sizes.large};
    `};
  `}
`
