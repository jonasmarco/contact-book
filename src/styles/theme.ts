export default {
  grid: {
    container: '120rem',
    defaultPadding: '1rem',
    gutter: '1.5rem'
  },
  border: {
    radius: '1rem',
    rounded: '5rem'
  },
  fonts: {
    sizes: {
      small: '1.2rem',
      normal: '1.8rem',
      large: '2.2rem',
      extraLarge: '2.8rem',
      bigger: '4.2rem'
    },
    weights: {
      slim: 300,
      normal: 400,
      bold: 600,
      extraBold: 700
    }
  },
  colors: {
    black: '#02020A',
    white: '#FCFCFC',
    red: '#FF6565',
    blue: '#4FB6FF',
    darkBlue: '#273549',
    gray: '#596E7D',
    yellow: '#F5A623',
    green: '#33CC33',

    primary: '#4FB6FF',
    secondary: '#596E7D',
    tertiary: '#30425C'
  },
  layout: {
    mobile: {
      gap: '2rem'
    },
    tablet: {
      gap: '6rem'
    },
    notebook: {
      gap: '10rem'
    },
    gapForm: '1rem'
  },
  text: {
    letterSpacing: '0.135rem',
    lineHeight: '2.8rem'
  },
  margin: {
    normal: '1.8rem',
    textMargin: '1.2rem'
  },
  layers: {
    base: 0,
    behindMenu: 5,
    menu: 10,
    modal: 100
  }
} as const
