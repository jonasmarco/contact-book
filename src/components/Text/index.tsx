import React from 'react'
import * as S from './styles'

export type Props = {
  children: string | React.ReactNode
  danger?: boolean
}

const Text = ({ children, danger = false }: Props) => {
  return <S.Wrapper danger={danger}>{children}</S.Wrapper>
}

export default Text
