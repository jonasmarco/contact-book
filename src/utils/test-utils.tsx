import { ReactElement } from 'react'
import { render } from '@testing-library/react'
import { BrowserRouter as Router } from 'react-router-dom'

import { ThemeProvider } from 'styled-components'
import theme from '../styles/theme'

const customRender = (ui: ReactElement) =>
  render(
    <Router>
      <ThemeProvider theme={theme}>{ui}</ThemeProvider>
    </Router>
  )

// re-export everything
export * from '@testing-library/react'

// override render method
export { customRender as render }
