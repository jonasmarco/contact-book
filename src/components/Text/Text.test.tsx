import { render } from '../../utils/test-utils'

import theme from '@/styles/theme'
import Text from '.'

describe('Text Component', () => {
  // (renderiza sem erros):
  // ensures that the Text is rendered without errors
  test('is rendered without error', () => {
    render(<Text>Lorem ipsum</Text>)
  })

  // (renderiza com o texto correto):
  // ensures that the Text is rendered with the correct text
  test('is rendered with correct text', () => {
    const { getByText } = render(<Text>Lorem ipsum</Text>)
    const textElement = getByText('Lorem ipsum')
    expect(textElement).toBeInTheDocument()
  })

  // (renderiza com o estilo correto)
  // ensures that the Text is rendered with the correct styles
  test('applies the correct default styles', () => {
    const { getByText } = render(<Text>Lorem ipsum</Text>)
    const textElement = getByText('Lorem ipsum')
    expect(textElement).toHaveStyle(`
      color: ${theme.colors.white};
      font-size: ${theme.fonts.sizes.small};
      line-height: ${theme.text.lineHeight};
      margin-bottom: ${theme.margin.textMargin};
    `)
  })

  // (renderiza com o estilo danger)
  // ensures that Text is rendered with the correct "danger" styles
  test('applies the correct danger styles', () => {
    const { getByText } = render(<Text danger>Lorem ipsum</Text>)
    const textElement = getByText('Lorem ipsum')
    expect(textElement).toHaveStyle(`
      color: ${theme.colors.red};
      font-size: ${theme.fonts.sizes.small};
      line-height: ${theme.text.lineHeight};
      margin-bottom: ${theme.margin.textMargin};
    `)
  })

  // (renderiza com o estilo correto em diferentes resoluções):
  // ensure that the Title is rendered with the correct styles according to the resolution
  test('applies different styles for tablet and notebook screens', () => {
    const { getByText } = render(<Text danger>Lorem ipsum</Text>)
    const textElement = getByText('Lorem ipsum')
    expect(textElement).toHaveStyleRule('font-size', theme.fonts.sizes.normal, {
      media: '(min-width: 769px)'
    })
    expect(textElement).toHaveStyleRule('font-size', theme.fonts.sizes.large, {
      media: '(min-width: 1367px)'
    })
  })

  // (snapshot)
  test('matches the snapshot', () => {
    const { asFragment } = render(<Text danger>Lorem ipsum</Text>)
    expect(asFragment()).toMatchSnapshot()
  })
})
