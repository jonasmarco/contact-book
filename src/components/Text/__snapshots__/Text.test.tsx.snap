// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text Component matches the snapshot 1`] = `
<DocumentFragment>
  .c0 {
  color: #FF6565;
  font-size: 1.2rem;
  line-height: 2.8rem;
  margin-bottom: 1.2rem;
}

@media (min-width:769px) {
  .c0 {
    font-size: 1.8rem;
  }
}

@media (min-width:1367px) {
  .c0 {
    font-size: 2.2rem;
  }
}

<p
    class="c0"
  >
    Lorem ipsum
  </p>
</DocumentFragment>
`;
